substrate/frame/revive/uapi/src/lib.rs

```diff
@@ -116,6 +116,27 @@ pub struct ReturnCode(u32);
 /// make sense for a memory location or length.
 const SENTINEL: u32 = u32::MAX;
 
+/// Converts a `u64` to a 32-byte array, encoded little-endian.
+///
+/// The `u64` value will be placed in the first 8 bytes,
+/// and the remaining bytes are set to zero.
+const fn u64_to_u256_bytes(value: u64) -> [u8; 32] {
+	let mut bytes = [0u8; 32];
+	let value_bytes = value.to_le_bytes();
+	let mut i = 0;
+	while i < 8 {
+		bytes[i] = value_bytes[i];
+		i += 1;
+	}
+	bytes
+}
+
+/// A constant representing `u64::MAX` as a 32-byte array, in little-endian encoding.
+///
+/// Deposit limits are `U256`, but balances are represented as `u64`.
+/// To represent no deposit limits on an operation, this should be used.
+pub const U256_MAX: [u8; 32] = u64_to_u256_bytes(u64::MAX);
```

Comment:
Should be the same, right?

```suggestion
pub const U256_MAX: [u8; 32] = [1; 32];
```

Reply:
This is going to raise [`pallet_revive::Error::BalanceConversionFailed`](https://github.com/paritytech/polkadot-sdk/blob/evm-contracts-limit-nested-calls-gas/substrate/frame/revive/src/lib.rs#L551).

I mistakenly assumed [`type Balance = u64`](https://github.com/paritytech/polkadot-sdk/blob/evm-contracts-limit-nested-calls-gas/substrate/frame/revive/src/lib.rs#L311C3-L311C22) always golds, and you're right that I [shouldn't](https://github.com/paritytech/polkadot-sdk/blob/evm-contracts-limit-nested-calls-gas/substrate/frame/revive/rpc/src/client.rs#L76).
However, in practice, is `Balance` ever made to be `U256` for the conversion from `[1; 32]` to not fail?

```
[1; 32] 
=
0x0101010101010101010101010101010101010101010101010101010101010101
=
2^248 + 2^240 + 2^232 + ... + 2^8 + 2^0
=
(256^32 - 1) / 255
=
115792089237316195423570985008687907853269984665640564039457584007913129639937
=
454086624460063511464984254936031011189294057512315937409637584344757371137
>>
340282366920938463463374607431768211455
=
U128::MAX
```

Reply:
Ahh okay you are right I forgot about that we are converting to the frame native type before we are capping the value.  To fix this you should change the code when converting the `deposit_limit` to use an `saturating_into`.  Then the contract can always use U256::MAX as it is unaware of the runtimes internal type.

> is Balance ever made to be U256 for the conversion from [1; 32] to not fail?

In practice it is `u64` in test runtimes and `u128` in real runtimes. But never `U256`.